src/Interpreters/InterpreterCreateQuery.cpp

```diff
@@ -1000,6 +1008,38 @@ void InterpreterCreateQuery::setEngine(ASTCreateQuery & create) const
         /// Some part of storage definition (such as PARTITION BY) is specified, but ENGINE is not: just set default one.
         if (!create.storage->engine)
             setDefaultTableEngine(*create.storage, getContext()->getSettingsRef().default_table_engine.value);
+        /// For exrternal tables with restore_replace_external_engine_to_null setting we replace external engines to
+        /// Null table engine.
+        else if (create.storage->engine->name == "AzureBlobStorage" ||
+            create.storage->engine->name == "AzureQueue" ||
+            create.storage->engine->name == "COSN" ||
+            create.storage->engine->name == "DeltaLake" ||
+            create.storage->engine->name == "Dictionary" ||
+            create.storage->engine->name == "Executable" ||
+            create.storage->engine->name == "ExecutablePool" ||
+            create.storage->engine->name == "ExternalDistributed" ||
+            create.storage->engine->name == "File" ||
+            create.storage->engine->name == "Hudi" ||
+            create.storage->engine->name == "Iceberg" ||
+            create.storage->engine->name == "JDBC" ||
+            create.storage->engine->name == "Kafka" ||
+            create.storage->engine->name == "MaterializedPostgreSQL" ||
+            create.storage->engine->name == "MongoDB" ||
+            create.storage->engine->name == "MySQL" ||
+            create.storage->engine->name == "NATS" ||
+            create.storage->engine->name == "ODBC" ||
+            create.storage->engine->name == "OSS" ||
+            create.storage->engine->name == "PostgreSQL" ||
+            create.storage->engine->name == "RabbitMQ" ||
+            create.storage->engine->name == "Redis" ||
+            create.storage->engine->name == "S3" ||
+            create.storage->engine->name == "S3Queue" ||
+            create.storage->engine->name == "TinyLog" ||
+            create.storage->engine->name == "URL")
+        {
+            if (getContext()->getSettingsRef().restore_replace_external_engine_to_null)
+                setNullTableEngine(*create.storage);
+        }
```

Comment:
I think the first check should be the setting `restore_replace_external_engine_to_null` to short-circuit the condition ASAP. Then, rather than having so many `or`s, I find it easier to understand and less error prone to have as a static/anonymous namespace at the beginning of this file something where you specify all of them:

```c++
static constexpr std::unordered_set<String> external_engines_to_null{"AzureQueue", "COSN", "DeltaLake", ...};
```

And then you do your check:

```c++
else if (getContext()->getSettingsRef().restore_replace_external_engine_to_null && external_engines_to_null.contains(create.storage->engine->name))
{
    setNullTableEngine(*create.storage);
}
```

Reply:
Thanks, that was more a draft attempt. I was writing a test and found out that this type of CREATE TABLE is not covered so more work needed in less convenient place.
```
(query: CREATE TABLE IF NOT EXISTS replicated2.mysql_schema_inference_function UUID '90be0f9f-7395-4d2c-a450-2fda46d3301e' (`id` Int32, `data` FixedString(16)) AS m
ysql('mysql80:3306', 'clickhouse', 'inference_table', 'root', '[HIDDEN]'), backup restore: false)
```

Reply:
Rewritten completely using information available in factory